Add support for process functions in `verdi plugin list`

sphuber opened the pull request:
Fixes #4115 

So far, only actual `Process` subclasses were supported, however,
process functions can be turned into a `Process` subclass just as well.
The wrapped function gets a new attribute `spec`, which when called
returns the `ProcessSpec` that is dynamically built based on the
function signature of the decorated function.

P.S.: this does not yet actually include a test through `verdi plugin list` for a registered process function since we don't have one yet. However, @mbercx should be working on a PR that would add one. If that get's merged before, I can use that to add an additional test.

sphuber commented:
@unkcpz here is another PR that could be suitable for you. There is no rush on this one. So if you are too busy now that is fine too.

unkcpz commented:
Why a CI test is failed here? I guess it has nothing to do with the current PR right?

sphuber commented on aiida/engine/processes/functions.py:
```diff
@@ -181,11 +181,15 @@ def decorated_function(*args, **kwargs):
         decorated_function.run_get_node = run_get_node
         decorated_function.is_process_function = True
         decorated_function.node_class = node_class
+        decorated_function.spec = spec
```
Good point! There is no reason :smile:

sphuber commented on aiida/cmdline/commands/cmd_plugin.py:
```diff
@@ -48,11 +49,11 @@ def plugin_list(entry_point_group, entry_point):
                 else:
                     echo.echo(str(plugin.get_description()))
-            except (AttributeError, TypeError):
+            except (AttributeError, TypeError) as exception:
                 echo.echo_error('No description available for {}'.format(entry_point))
```
>     * I guess `TypeError` here is used to catch the error when plugin is a function, since it is already handled, can be removed.

Note that this can be called also for objects that are neither a process class nor a process function, because this function is used for all entry point groups. But I guess your point still stands, I think for any object, it should ultimately just raise `AttributeError` when we call `plugin.is_process_function`, so I think we can safely just catch that. I will update it.

>     * The `exception` here not used, can we just get rid of it or print it in error message?

Yes, I will remove it.

sphuber commented:
> Why a CI test is failed here? I guess it has nothing to do with the current PR right?

Looks like a random transient problem and should be fixed when we rerun

sphuber commented:
> I just wondering is there any particular reason you create a new fixture for cli testing? If so, is that all cmdline cli run test should be transformed to this strategy?

The reason I added this is because I think it makes the typical CLI tests written in pytest really concise and clear. It does all the logic that the current tests typically do, such as check for exception, and get output in lines with whitespace stripped. Currently, all tests are doing all these things a bit ad-hoc and makes the code a bit inconsistent. Ideally we would convert all tests like that to use this fixture, but don't want to do it in this PR. Anyway, the majority of tests so far still use unittest instead of `pytest`.

unkcpz commented on tests/cmdline/commands/test_plugin.py:
```diff
@@ -0,0 +1,43 @@
+    entry_point = factory(entry_point_name)
+
+    result = run_cli_command(cmd_plugin.plugin_list, [entry_point_group, entry_point_name])
+    assert entry_point.__doc__ in result.output
```
Shouldn't these two be exactly equal?

sphuber replied:
no, because the output of the command adds a lot of other things such as a formatted version of the spec

unkcpz commented:
> > I just wondering is there any particular reason you create a new fixture for cli testing? If so, is that all cmdline cli run test should be transformed to this strategy?
> 
> The reason I added this is because I think it makes the typical CLI tests written in pytest really concise and clear. It does all the logic that the current tests typically do, such as check for exception, and get output in lines with whitespace stripped. Currently, all tests are doing all these things a bit ad-hoc and makes the code a bit inconsistent. Ideally we would convert all tests like that to use this fixture, but don't want to do it in this PR. Anyway, the majority of tests so far still use unittest instead of `pytest`.

I see, I recomend to create a issue for future improvement. And since the `raise` part of this fixture is not being tested, would be better to first include that kinds of tests.

sphuber commented on tests/cmdline/commands/conftest.py:
```diff
@@ -0,0 +1,35 @@
+            assert result.exception is None, ''.join(traceback.format_exception(*result.exc_info))
+            assert result.exit_code == 0, result.output
+
+        result.output_lines = [line.strip() for line in result.output.split('\n') if line.strip()]
```
Ha! It is actually just an attribute that I add at that very line. The actual `Result` class provides `output` and `output_bytes` but I find I often need the list of stripped lines, so as a shortcut I added it here as an attribute

sphuber commented:
Thanks for the second review @unkcpz , I have updated the `run_cli_command` fixture and added a test that uses the `raises` argument. If you approve this PR, I will open the issue to start replacing existing tests with this new fixture and link it here.

unkcpz commented:
Thanks! @sphuber

unkcpz commented:
As a reviewer, I guess it is also my duty to close https://github.com/aiidateam/aiida-core/issues/4115 ?

sphuber commented:
> As a reviewer, I guess it is also my duty to close #4115 ?

No, because I have marked my OP that this closes that issue and upon merging into `develop` Github will automatically close it

sphuber commented:
Thanks a lot for the reviews @unkcpz

sphuber commented:
Opened #4142 to address the refactoring of existing CLI tests to use the new fixture
